src/sql/relation.rs

```diff
@@ -307,16 +311,35 @@ impl<'a, T: QueryToRelationTranslator + Copy + Clone> VisitedQueryRelations<'a,
                             .into_iter()
                             .filter_map(|f| join.right().schema().field(f.name()).is_ok().then_some(f.name().to_string()))
                             .collect();
-                        join.remove_duplicates_and_coalesce(v,&new_columns)
+                        join.remove_duplicates_and_coalesce(v, &join_columns, preserve_input_names)
                     },
                     ast::JoinOperator::LeftSemi(_) => todo!(),
                     ast::JoinOperator::RightSemi(_) => todo!(),
                     ast::JoinOperator::LeftAnti(_) => todo!(),
                     ast::JoinOperator::RightAnti(_) => todo!(),
-                    _ => Relation::from(join),
+                    _ => join.remove_duplicates_and_coalesce(vec![], &join_columns, preserve_input_names),
+                };
+
+                
+                let composed_columns = if preserve_input_names {
+                    // relation fields are renamed to those of the join relation inputs
+                    // if no name collision is generated.  
+                    let original_not_ambiguous_columns: Hierarchy<Identifier> = join_columns
+                        .iter()
+                        .map(|(key, value)| {
+                            let original_col_name = key.last().unwrap().as_str();
+                            let join_col_name = value.head().unwrap();
+                            match join_columns.get(&[original_col_name.to_string()]) {
+                                Some(_) => (Identifier::from(join_col_name), original_col_name.into()),
+                                None => (Identifier::from(join_col_name), join_col_name.into()),
+                            }
+                    })
+                    .collect();
+                    composed_columns.and_then(original_not_ambiguous_columns)
+                } else {
+                    composed_columns
```

Comment:
You're relying on the previously defined composed column here ? It's a bit hard to read